rc/tools/doc.kak

```diff
@@ -154,14 +157,94 @@ define-command -params 1..2 \
     evaluate-commands %sh{
         readonly page="${kak_runtime}/doc/${1}.asciidoc"
         if [ -f "${page}" ]; then
+            printf 'evaluate-commands -try-client %%opt{docsclient} %%{
+                edit! -scratch "*doc-%s*"
+                doc-render "%s"
+            }' "$(basename "$1" .asciidoc)" "${page}"
             if [ $# -eq 2 ]; then
-                jump_cmd="doc-jump-to-anchor '$2'"
+                printf "doc-jump-to-anchor '%s'" "$2"
             fi
-            printf %s\\n "evaluate-commands -try-client %opt{docsclient} %{ doc-render ${page}; ${jump_cmd} }"
         else
             printf 'fail No such doc file: %s\n' "${page}"
         fi
     }
 }
 
 alias global help doc
+
+define-command -params 1.. -docstring %{
+        doc-search <pattern> [topic...]: search for a pattern in the documentation
+
+        Display a menu listing all the search matches in the given topics, or the whole documentation if unspecified (the number of search results is effected by the `doc_max_search_results` option)
+    } \
+    -shell-script-candidates %{
+        case "${kak_token_to_complete}" in
+            0) ;;
+            *) find "${kak_runtime}/doc/" -type f -name '*\.asciidoc' | sed 's,.*/,,; s,\.[^/]*$,,';;
+        esac
+    } doc-search %{
+    set-option global doc_search_matches
+
+    evaluate-commands %sh{
+        readonly SEARCH_PATTERN="${1}"
+
+        shift
+        if [ $# -eq 0 ]; then
+            eval set -- $(ls "${kak_runtime}/doc"/*.asciidoc | sed 's,.*/,,; s,\.[^/]*$,,')
+        else
+            for topic; do
+                if [ ! -e "${kak_runtime}/doc/${topic}.asciidoc" ]; then
+                    printf 'fail No such topic %%{%s}' "${topic}"
+                    exit 1
+                fi
+            done
+        fi
+
+        for topic; do
+            SEARCH_PATTERN_ESC=$(printf %s "${SEARCH_PATTERN}" | sed "s/'/&&/g")
+            printf "doc-search-impl '%s' '%s'\n" "${topic}" "${SEARCH_PATTERN_ESC}"
+        done
+    }
+
+    evaluate-commands %sh{
+        kakquote() { printf "%s" "$*" | sed "s/'/''/g; 1s/^/'/; \$s/\$/'/"; }
+
+        eval "set -- ${kak_quoted_opt_doc_search_matches}"
+
+        while [ $# -gt 0 ]; do
+            candidates="${candidates} $(kakquote "{MenuInfo}${1}{default}{\\} ${3}") 'doc ${1}; select $2'"
+            shift 3
+        done
+
+        if [ -n "${candidates}" ]; then
+            printf '
+                set global doc_search_matches
+                menu -markup --%s
+            ' "${candidates}"
+        fi
+    }
+}
+
+define-command -hidden -params 2 doc-search-impl %{
+    edit -scratch
+
+    try %{
+        doc-render "%val{runtime}/doc/%arg{1}.asciidoc"
```

Comment:
It really is a bit slow. I wouldn't mind using a version that does not render the asciidoc. The rendering shouldn't make a difference for most searches' results.

Maybe just render the results of grep -R doc/. Not sure if that would be robust enough.

Reply:
We need to render the markup because we don't want to search through text that is part of the markup language, e.g. link.

Some markup is also replaced in the output, which invalidates coordinates if we want to jump to a specific location after a call to `:doc`.